src/search/landmarks/landmark_status_manager.cc

```diff
@@ -1,171 +1,223 @@
+BitsetView LandmarkStatusManager::get_past_landmarks(const State &state) {
+    return past_landmarks[state];
 }
 
-void LandmarkStatusManager::process_initial_state(
-    const State &initial_state, utils::LogProxy &log) {
-    set_reached_landmarks_for_initial_state(initial_state, log);
-    update_lm_status(initial_state);
+BitsetView LandmarkStatusManager::get_future_landmarks(const State &state) {
+    return future_landmarks[state];
 }
 
-void LandmarkStatusManager::set_reached_landmarks_for_initial_state(
-    const State &initial_state, utils::LogProxy &log) {
-    BitsetView reached = get_reached_landmarks(initial_state);
-    // This is necessary since the default is "true for all" (see comment above).
-    reached.reset();
-
-    int inserted = 0;
-    int num_goal_lms = 0;
-    for (auto &lm_node : lm_graph.get_nodes()) {
-        const Landmark &landmark = lm_node->get_landmark();
-        if (landmark.is_true_in_goal) {
-            ++num_goal_lms;
-        }
+ConstBitsetView LandmarkStatusManager::get_past_landmarks_const(const State &state) const {
+    return past_landmarks[state];
+}
 
-        if (!lm_node->parents.empty()) {
-            continue;
-        }
-        if (landmark.conjunctive) {
-            bool lm_true = true;
-            for (const FactPair &fact : landmark.facts) {
-                if (initial_state[fact.var].get_value() != fact.value) {
-                    lm_true = false;
-                    break;
-                }
-            }
-            if (lm_true) {
-                reached.set(lm_node->get_id());
-                ++inserted;
+ConstBitsetView LandmarkStatusManager::get_future_landmarks_const(const State &state) const {
+    return future_landmarks[state];
+}
+
+void LandmarkStatusManager::progress_initial_state(const State &initial_state) {
+    BitsetView past = get_past_landmarks(initial_state);
+    BitsetView fut = get_future_landmarks(initial_state);
+
+    for (auto &node : lm_graph.get_nodes()) {
+        int id = node->get_id();
+        const Landmark &lm = node->get_landmark();
+        if (lm.is_true_in_state(initial_state)) {
+            assert(past.test(id));
+            for (auto &parent : node->parents) {
+                /*
+                  We assume here that no natural orderings A->B are generated
+                  such that B holds in the initial state. Such orderings would
+                  only be valid in unsolvable problems.
+                */
+                assert(parent.second <= EdgeType::REASONABLE);
+                /*
+                  Reasonable orderings A->B for which both A and B hold in the
+                  initial state are immediately satisfied. We assume such
+                  orderings are not generated in the first place.
+                */
+                assert(!parent.first->get_landmark().is_true_in_state(
+                           initial_state));
+                utils::unused_variable(parent);
+                fut.set(id);
```

Comment:
Doesn't this loop waste a lot of time in non-debug mode if there are multiple such parents? I find it nonintuitive to have a loop (at least in non-debug mode) where the loop body doesn't depend on the loop variable, even when this is not time-critical, and would perhaps prefer doing the loop purely for the debug test if we want it separately from the fut.set.id() without a loop (instead testing parents for emptiness).